prql-compiler/src/sql/srq/ast.rs

```diff
@@ -0,0 +1,170 @@
+//! Sql Relational Query AST
+
+use anyhow::Result;
+use enum_as_inner::EnumAsInner;
+use itertools::Itertools;
+use serde::Serialize;
+
+use crate::ast::pl::{ColumnSort, InterpolateItem, JoinSide, RelationLiteral};
+use crate::ast::rq::{self, Expr, RqFold, TId};
+
+#[derive(Debug, Clone)]
+pub struct SqlQuery {
+    /// Common Table Expression (WITH clause)
+    pub ctes: Vec<Cte>,
+
+    /// The body of SELECT query.
+    pub main_relation: SqlRelation,
+}
+
+#[derive(Debug, Clone, EnumAsInner, Serialize)]
+pub enum SqlRelation {
+    AtomicPipeline(Vec<SqlTransform<RelationExpr, ()>>),
+    Literal(RelationLiteral),
+    SString(Vec<InterpolateItem<Expr>>),
+}
+
+#[derive(Debug, Clone, Serialize)]
+pub enum RelationExpr {
+    Ref(TId, Option<String>),
+    SubQuery(SqlRelation, Option<String>),
+}
+
+#[derive(Debug, Clone)]
+pub struct Cte {
+    pub tid: TId,
+    pub kind: CteKind,
+}
+
+#[derive(Debug, Clone)]
+pub enum CteKind {
+    Normal(SqlRelation),
+    Loop {
+        initial: SqlRelation,
+        step: SqlRelation,
+        recursive_name: String,
+    },
+}
+
+/// Similar to [rq::Transform], but more similar to SQL clauses.
+///
+/// Uses a two generic args that allows compiler to do work in multiple stages.
+/// First convert RQ to [SqlTransform<TableRef, rq::Transform>] and at the end
+/// compile that to [SqlTransform<RelationExpr, ()>].
```

Comment:
Thanks for explaining. I'm probably 70% of the way to having a good mental model of this